### ISSUE TITLE
sysctl dependency change problem

### ISSUE THREAD

**Issue by buzzdeee** [xaque208/puppet-bsd#90]
Since I run all my moudles from upstream git, and update daily, I 
kind of quickly ran into this issue here ;)

I have duritong/sysctl installed in parallel to hercules/augeasproviders_sysctl
and I get with latest git:
# puppet agent --test

Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Invalid parameter value on Sysctl[net.inet.ip.forwarding] at /etc/puppet/environments/production/modules/bsd/manifests/network.pp:31 on node MYNODE
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run

git reset --hard 39d22efac07fd69853f3783550ca06b21584b955

then I at least only get:
# puppet agent --test

Info: Retrieving pluginfacts
Info: Retrieving plugin
Notice: /File[/var/puppet/lib/puppet/provider/sysctl/augeas.rb]/ensure: defined content as '{md5}227da37dd9052c3bbb4bc53f69936afb'
Info: Loading facts
Info: Loading facts
Info: Caching catalog for MYNODE
Warning: Found multiple default providers for sysctl: augeas, parsed; using augeas
Info: Applying configuration version '1466582508'

luckily I use duritong/sysctl only in my own
modules and can fix that, but others might run into similar dependency issues.

**Comment by xaque208** [comment]
Ah, thank you for the report!  I'll double check those parameters and work to fix, probably tonight.

**Comment by xaque208** [comment]
Alright, so lunch breaks work for testing too.  This is odd.  Does your system's versions match the failure case in the travis build here? https://github.com/xaque208/puppet-bsd/pull/91

**Comment by buzzdeee** [comment]
nope, I'm running snapshots, (more or less with my puppetmaster), but I have there ruby 2.2.5 and puppet 3.8.7. meanwhile I updated my three modules that used to use duritong/sysctl, and I'm now back in business with the augeas one.

**Comment by xaque208** [comment]
Well thank you for the report all the same.  I failed to correctly estimate the impact of using both sysctl modules.  I'll work on some tests to help the situation and get this sorted out, likely with a new major version.  Sorry for the trouble.  So you are all up to date on this module an functioning as expected?  Also as an aside, are you planing to make the upgrade to 4.x?

**Comment by buzzdeee** [comment]
yep, I updated all my modules to use the augeas based sysctl module, was not a too big deal.
That works well for me here, no issue so far.

With regard to Puppet 4, don't know how close you follow -current. Puppet 4 is enabled in the snapshots, kind of latest Puppetdb, facter 3, hiera etc.

I run my puppetmaster as gunicorn proxies behind nginx, since recently, using policy based certificate signing, see PUP-6366 ;) I also make heavy use of hiera, with eyaml and file backends.
I've a test puppetmaster setup on a notebook. Luckily I use future parser more or less since the beginning, as well as structured facts since some time.
However, some problems, esp. with the two hiera backends:
 - their packages are requiring hiera1, which conflicts with puppet4, I can get away with it deploying just the ruby gems, easy
 - with hiera-file I use dot's in the keys that I look up, but that is "borked" with hiera 3 now, since the . is special now, because of these structured facts :(, basically I ran into this issue: https://github.com/voxpupuli/hiera-file/issues/13, I can get away with that, changing all the file names to have underscores or the like instead of the dots, just a bit of work, I stuffed quite a bunch of files into my hierarchy ;)

last but not least, I need to find a way to get Puppet 3 agents working behind Puppet 4 master, i.e. a proper upgrade path. For me @home no real issue, it would not matter, I'd just re-deploy my puppetmaster with Puppet4, and then re-deploy my nodes and being done with it, I'm installing new snapshots every couple of days/weeks on my nodes anyways. but @work, that won't work out easily. A mixture of OS/Versions running the agents.
So, I found this here:
https://docs.puppet.com/puppetserver/latest/compatibility_with_puppet_agent.html
but haven't yet got around trying to teach nginx how to properly redirect Puppet3 API to Puppet4 API puppetmaster. The document only describes tweaks to the puppet-server, but that is not yet quite working on OpenBSD. If you have any hint/cheat sheet to get that to work for gunicorn/nginx, I would be grateful ;)

**Comment by xaque208** [comment]
I've been running purely puppet4 for about the last year.  My employer made this easy as most recently I deployed a new infrastructure using newer puppet, not not much to do there but deploy puppetserver, puppetdb, and puppet4.

I only have BSD at home though, and strictly BSD except for a couple RPis, so to upgrade those systems I used a combination of solutions.  On systems that don't have up to date packages, I decided to use the gem instead of the native package.  On the raspberry pi etc, the ARM platform doesn't have puppet-agent packages to install until recently, so the gem works well enough for those cases, and got me out of waiting for Puppet to ship updated packages.  I took the same approach for OpenBSD as well, which was a bit more work due to the links not being created in `/usr/local/bin`, but now its working well enough for me to convert all of my code to puppet4 without waiting on OpenBSD to ship a puppet4 package.  For the FreeBSD systems I have, I maintain the `sysutils/puppet4` port, and related `databases/puppetdb3` and `sysutils/puppetserver`, so thats been more focus for me, developing, testing, and submitting those ports to the FreeBSD tree.

I mostly just got tired of waiting for packages to update, as I want(ed) to use newer language features and keep up to date, both in knowledge and code.  When packages come around, I'll use those instead of the gem, but the gem is a fine stopgap.

For most of my modules, I've been updating them to use new features that are not supported in 3.x, so most of the puppet modules I maintain are moving away from supporting puppet 3 entirely.  I think this BSD module is one of the last to support puppet 3.x on its latest version, and I'd like to change that soon.  If you are tracking master, this will end up being a problem for you, as master will get less supported on 3.x until its just not.  I try to keep good tags, so tracking those might be more stable, though having you track master was good since you were able to raise this issue :).

I can't offer much advice on the gunicorn stuff, other than to say that the effort there might be better spend getting over to puppteserver.  I think there is a winning combo of puppetserver version + puppetdb version + puppet 3.x that will work, but I don't know what those versions are.

I'd be happy enough to share the code I have for managing OpenBSD with a gem if thats useful.  A bit long winded, but maybe useful.

Cheers
